kerneltest/kerneltest.sh

```diff
@@ -0,0 +1,127 @@
+}
+
+build_test() {
+    echo "building test"
+    CGO_ENABLED=1 GOOS=linux GOARCH=amd64 CC="clang" CGO_CFLAGS="-I/home/javierhonduco/code/parca-agent/dist/libbpf/amd64/usr/include" CGO_LDFLAGS="-fuse-ld=ld /home/javierhonduco/code/parca-agent/dist/libbpf/amd64/libbpf.a" go test -v ./pkg/profiler/cpu -c -ldflags='-extldflags=-static' -trimpath -tags 'osusergo netgo'
+    mv cpu.test kerneltest/
+}
+
+create_initramfs() {
+    echo "creating initramfs"
+    bluebox -e kerneltest/cpu.test
+    mv initramfs.cpio kerneltest
+}
+
+# TODO: host the kernels ourselves to not use cilium's quotas
+download_kernel() {
+    kernel_version=$1
+    echo "downloading kernel $kernel_version"
+    curl -o "kerneltest/kernels/linux-$kernel_version.bz" -s -L -O --fail "https://github.com/cilium/ci-kernels/raw/master/linux-$kernel_version.bz"
+}
+
+use_kernel() {
+    kernel_version=$1
+    if [ ! -f "kerneltest/kernels/linux-$kernel_version.bz" ]; then
+        echo "kernel $kernel_version not found"
+        download_kernel "$kernel_version"
+    fi
+}
+
+github_start() {
+    kernel_version=$1
+    [[ -z "${GITHUB_ACTIONS}" ]] || echo "::group:: running tests on kernel $kernel_version"
+}
+
+github_end() {
+    kernel_version=$1
+    [[ -z "${GITHUB_ACTIONS}" ]] || echo "::endgroup::"
+}
+
+run_tests() {
+    kernel_version=$1
+    memory=$2
+    echo "running tests in qemu"
+    github_start "$kernel_version"
+    # kernel.panic=-1 and -no-reboot ensures we won't get stuck on kernel panic.
+    # TODO: add the ability to not print anything so we can integrate with make and do something like:
+    # vmtests:
+    # 	NO_TTY=1 ./kerneltest/kerneltest.sh
+    qemu-system-x86_64 -no-reboot -append 'printk.devkmsg=on kernel.panic=-1 crashkernel=256M' \
+        -nographic -append "console=ttyS0" -m "$memory" -kernel "kerneltest/kernels/linux-$kernel_version.bz" \
+        -initrd kerneltest/initramfs.cpio | tee "kerneltest/logs/vm_log_$kernel_version.txt"
+    github_end "$kernel_version"
+}
+
+did_test_pass() {
+    kernel_version=$1
+    grep PASS "kerneltest/logs/vm_log_$kernel_version.txt" >/dev/null
+}
+
+check_executable() {
+    executable=$1
+    if ! command -v "$executable" &>/dev/null; then
+        echo "$executable could not be found"
+        exit 1
+    fi
+}
+
+# Initial checks
+check_executable "curl"
+check_executable "qemu-system-x86_64"
+check_executable "bluebox"
+
+# Build artifacts for testing.
+build_bpf
+build_test
+create_initramfs
+
+# Run the tests.
+kernel_versions=("5.10" "5.18" "5.19")
+
+for kernel in "${kernel_versions[@]}"; do
+    use_kernel "$kernel"
+    run_tests "$kernel" "1.5G"
+done
+
+# TODO: return the proper status code
+echo "============="
+echo "Test results:"
+echo "============="
+for kernel in "${kernel_versions[@]}"; do
+    if did_test_pass "$kernel"; then
+        echo "- ✅ $kernel"
+    else
+        echo "- ❌ $kernel"
+    fi
+done
+
+# BUG: we need to check for batch operation support and don't run these tests.
```

Comment:
If we plan on using bach ops, we can add some detection and not run the tests, otherwise, perhaps we can move this to another test? cc @v-thakkar @kakkoyun

Reply:
Better to run in a dedicated pipeline.

Reply:
Hiii! What do you mean by pipeline? 

I've added some basic feature checks right now. Ideally, in the future, we would add this check to the `profiler` struct + wrap the error with `%w` in libbpfgo so the error can be unwrapped. Going with the current approach as we don't use batch operations anywhere yet, and changing the error to be wrapped in libbpfgo would take some days to land.

Reply:
Hi 👋 What I meant is to run kernel tests as in their own GitHub action, not part of the `build.yaml`. It would be easier to track down the failures and how long do they take.